c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -1905,7 +2069,37 @@ void raw_delete(void* ptr) {
   caching_allocator.free(ptr);
 }
 
+std::vector<std::vector<AllocFreeEvent>> getAllocFreeEvents() {
+  return memory_tracker.get_alloc_free_events();
+}
+
+void set_mem_plan(std::vector<std::vector<AllocFreeEvent>> plan) {
+  std::vector<Block*> plan_ptrs{plan.size(), nullptr};
+  std::vector<size_t> plan_size{plan.size(), 0};
+  std::vector<unsigned int> non_rec_lengths{plan.size(), 0};
+  void* r;
+  for (unsigned int device = 0; device < plan.size(); device++) {
+    plan_size[device] = plan[device][0].size;
+    non_rec_lengths[device] = plan[device][0].ptr;
+    r = nullptr; // if ptr remains null: this means no planning for this device
+                 // because either device memory is not planned or failed to
+                 // assign enough memory for planning
```

Comment:
> failed to assign enough memory for planning

this will throw an OOM error as usual? Please add tests for it.

Reply:
I changed this part. Thank you.